Fireworks/model.py

```diff
@@ -0,0 +1,192 @@
+        for key in self.required_components:
+            if key not in components:
+                missing_components.append(key)
+                error = True
+        if error:
+            raise ParameterizationError("Missing required components {0}".format(missing_components))
+
+    @property
+    def required_components(self):
+        """
+        This should be overridden by a subclass in order to specify components that should be provided during initialization. Otherwise,
+        this will default to just return the components already present within the Model.
+        """
+        return self.components
+
+    @abstractmethod
+    def forward(self, message):
+        """
+        Represents a forward pass application of the model to an input. Must be implemented by a subclass.
+        This should return a Message.
+        """
+        pass
+
+    def _change_temperature(self, boo, components = None):
+
+        self.update_components() # This is here so that the model checks to see if Parameters/Modules have changed before freezing/unfreezing
+
+        if isinstance(components, str): # eg. a single component is provided as a string
+            components = [components]
+
+        if components is None:
+            components = self.components
+
+        for component in components:
+            if isinstance(getattr(self, component), Parameter):
+                getattr(self, component).requires_grad = boo
+            elif isinstance(getattr(self, component), Model):
+                getattr(self, component)._change_temperature(boo) # Recursively freezes Models
+            elif isinstance(getattr(self, component), Module): # Is a PyTorch module but not a model.
+                _change_temperature(boo, getattr(self, component))
+
+    def freeze(self, components = None):
+        """
+        Freezes the given components of the model (or all of them if none are specified) in order to prevent gradient updates.
+        This means setting requires_grad to false for specified components so that these components
+        are not updated during training.
+        """
+        self._change_temperature(False, components)
+
+    def unfreeze(self, components = None):
+        """
+        Unfreezes the given components of the model (or all of them if none are specified) in order to prevent gradient updates.
+        This means setting requires_grad to true for specified components so that these components
+        are updated during training.
+        """
+        self._change_temperature(True, components)
+
+    # TODO: Figure out model i/o
+    # TODO: Implement description methods
+
+    def __getitem_hook(self, message):
+
+        return self.forward(message)
+
+    def __next_hook(self, message):
+
+        return self.forward(message)
+
+    def __call__(self, message, *args, **kwargs):
+
+        try: # This will trigger a recursive call if possible.
+            message = self.input(message, *args, **kwargs)
```

Comment:
I think from a readability standpoint define this with a different variable like `recursive_message` or something. You should also probably define this at the beginning with `None` and maybe do a type check to make it more clear